File: bayeurs-backend/src/services/commentService.ts
import { query } from '../config/database.js';

export class CommentService {
  static async postComment(userId: string, propertyId: string, content: string, rating?: number, parentId?: string) {
    const sql = `
      INSERT INTO comments (user_id, property_id, content, rating, parent_id)
      VALUES ($1, $2, $3, $4, $5)
      RETURNING *;
    `;
    const result = await query(sql, [userId, propertyId, content, rating, parentId]);
    return result.rows[0];
  }

  static async getPropertyComments(propertyId: string, limit: number = 20, offset: number = 0) {
    const sql = `
      WITH RECURSIVE comment_tree AS (
        -- Root comments
        SELECT c.*, u.username, 0 as level
        FROM comments c
        JOIN users u ON c.user_id = u.id
        WHERE c.property_id = $1 AND c.parent_id IS NULL
        
        UNION ALL
        
        -- Replies
        SELECT c.*, u.username, ct.level + 1
        FROM comments c
        JOIN users u ON c.user_id = u.id
        JOIN comment_tree ct ON c.parent_id = ct.id
      )
      SELECT * FROM comment_tree
      ORDER BY created_at ASC
      LIMIT $2 OFFSET $3;
    `;
    const result = await query(sql, [propertyId, limit, offset]);
    return result.rows;
  }

  static async updateInteractions(commentId: string, type: 'like' | 'dislike') {
    const column = type === 'like' ? 'likes_count' : 'dislikes_count';
    const sql = `UPDATE comments SET ${column} = ${column} + 1 WHERE id = $1 RETURNING *;`;
    const result = await query(sql, [commentId]);
    return result.rows[0];
  }

  static async reportComment(commentId: string) {
    const sql = 'UPDATE comments SET is_reported = TRUE WHERE id = $1 RETURNING *;';
    const result = await query(sql, [commentId]);
    return result.rows[0];
  }

  static async deleteComment(commentId: string, userId: string) {
    const sql = 'DELETE FROM comments WHERE id = $1 AND user_id = $2 RETURNING id;';
    const result = await query(sql, [commentId, userId]);
    return result.rows.length > 0;
  }
}

File: app_mobil_bayeur/lib/services/comment_service.dart
import 'package:app_mobil_bayeur/models/comment_model.dart';
import 'package:app_mobil_bayeur/services/api_service.dart';

class CommentService {
  final ApiService _api;

  CommentService(this._api);

  Future<List<Comment>> getPropertyComments(String propertyId, {int limit = 20, int offset = 0}) async {
    final response = await _api.get('/comments/property/$propertyId', params: {
      'limit': limit,
      'offset': offset,
    });
    return (response.data as List).map((json) => Comment.fromJson(json)).toList();
  }

  Future<Comment> postComment(String propertyId, String content, {int? rating, String? parentId}) async {
    final response = await _api.post('/comments', data: {
      'property_id': propertyId,
      'content': content,
      'rating': rating,
      'parent_id': parentId,
    });
    return Comment.fromJson(response.data);
  }

  Future<void> interactWithComment(String commentId, String type) async {
    await _api.post('/comments/$commentId/interact', data: {'type': type});
  }

  Future<void> reportComment(String commentId) async {
    await _api.post('/comments/$commentId/report');
  }

  Future<void> deleteComment(String commentId) async {
    await _api.delete('/comments/$commentId');
  }
}
